src/NuGet.Core/NuGet.Packaging/PackageExtraction/ZipArchiveExtensions.cs

```diff
@@ -67,25 +68,65 @@ public static string SaveAsFile(this ZipArchiveEntry entry, string fileFullPath,
 
         public static void UpdateFileTimeFromEntry(this ZipArchiveEntry entry, string fileFullPath, ILogger logger)
         {
-            var attr = File.GetAttributes(fileFullPath);
+            Testable.Default.UpdateFileTimeFromEntry(entry, fileFullPath, logger);
+        }
+
+        internal class Testable
+        {
+            public static Testable Default { get; } = new Testable(EnvironmentVariableWrapper.Instance);
 
-            if (!attr.HasFlag(FileAttributes.Directory) &&
-                entry.LastWriteTime.DateTime != DateTime.MinValue && // Ignore invalid times
-                entry.LastWriteTime.UtcDateTime <= DateTime.UtcNow) // Ignore future times
+            internal Testable(IEnvironmentVariableReader environmentVariableReader)
             {
-                try
+                _updateFileTimeFromEntryMaxRetries = 9;
+                string value = environmentVariableReader.GetEnvironmentVariable("NUGET_UpdateFileTime_MaxRetries");
+                if (int.TryParse(value, out int maxRetries) && maxRetries > 0)
                 {
-                    File.SetLastWriteTimeUtc(fileFullPath, entry.LastWriteTime.Add(entry.LastWriteTime.Offset).UtcDateTime);
+                    _updateFileTimeFromEntryMaxRetries = maxRetries;
                 }
-                catch (ArgumentOutOfRangeException ex)
+            }
+
+            private readonly int _updateFileTimeFromEntryMaxRetries;
+
+            public void UpdateFileTimeFromEntry(ZipArchiveEntry entry, string fileFullPath, ILogger logger)
+            {
+                if (entry == null) throw new ArgumentNullException(nameof(entry));
+                if (fileFullPath == null) throw new ArgumentNullException(nameof(fileFullPath));
+                if (logger == null) throw new ArgumentNullException(nameof(logger));
+
+                var attr = File.GetAttributes(fileFullPath);
+
+                if (!attr.HasFlag(FileAttributes.Directory) &&
+                    entry.LastWriteTime.DateTime != DateTime.MinValue && // Ignore invalid times
+                    entry.LastWriteTime.UtcDateTime <= DateTime.UtcNow) // Ignore future times
                 {
-                    string message = string.Format(
-                        CultureInfo.CurrentCulture,
-                        Strings.FailedFileTime,
-                        fileFullPath, // {0}
-                        ex.Message); // {1}
+                    try
+                    {
+                        int retry = 0;
+                        bool successful = false;
+                        while (!successful)
+                        {
+                            try
+                            {
+                                File.SetLastWriteTimeUtc(fileFullPath, entry.LastWriteTime.Add(entry.LastWriteTime.Offset).UtcDateTime);
+                                successful = true;
```

Comment:
If under some circumstances (hypothetically AV suspect file has virus signature and quarantines) if we fail to set `successful` to `true` then this loop could never break even though `retry` >= `_updateFileTimeFromEntryMaxRetries` condition is already met. Maybe we need to add another short-circuit for above condition too.

Reply:
I don't follow. If `File.SetLastWriteTimeUtc` throws an `IOException` due to bad sector, etc, then it will retry up to `_updateFileTimeForEntryMaxRetries` times.

If it throws any other exception, the catch block in the while loop will not catch the exception, and the exception will bubble up the callstack, and therefore the while loop is aborted.

If `File.SetLastWriteTime` does't throw, we set `successful = true` and exit the loop.

I don't see any way for there to be an infinite loop.

Reply:
Sorry, I misunderstood about how `catch when` works, resolving.